artiq/coredevice/shuttler.py

```diff
@@ -89,3 +124,140 @@ def __init__(self, dmgr, channel, core_device="core"):
     @kernel
     def trigger(self, trig_out):
         rtio_output(self.target_o, trig_out)
+
+
+RELAY_SPI_CONFIG = (0*spi.SPI_OFFLINE | 1*spi.SPI_END |
+                    0*spi.SPI_INPUT | 0*spi.SPI_CS_POLARITY |
+                    0*spi.SPI_CLK_POLARITY | 0*spi.SPI_CLK_PHASE |
+                    0*spi.SPI_LSB_FIRST | 0*spi.SPI_HALF_DUPLEX)
+
+SPI_CS_RELAY = 1 << 0
+SPI_CS_LED = 1 << 1
+SPI_DIV = 4
+
+ADC_SPI_CONFIG = (0*spi.SPI_OFFLINE | 0*spi.SPI_END |
+                  0*spi.SPI_INPUT | 0*spi.SPI_CS_POLARITY |
+                  1*spi.SPI_CLK_POLARITY | 1*spi.SPI_CLK_PHASE |
+                  0*spi.SPI_LSB_FIRST | 0*spi.SPI_HALF_DUPLEX)
+
+ADC_CS = 1
+ADC_SPI_DIV = 32
```

Comment:
I haven't really looked into the proper clock division factor. Register writing is likely a lot faster.